src/gc/gc.d

```diff
@@ -2122,6 +2308,12 @@ struct Gcx
                 auto pool = findPool(p);
                 if (pool)
                 {
+                    static if (precise) if (p1pool)
+                    {
+                        size_t p1off = cast(size_t)(p1 - cast(void**)p1pool.baseAddr);
```

Comment:
You should move this test to before p1 is dereferenced. If something isn't a pointer we should perform as little work as possible. This loop is key to make this thing fast.